compiler/src/dotty/tools/dotc/core/ScalaFunction.scala

```diff
@@ -0,0 +1,122 @@
+package dotty.tools.dotc.core
+
+import dotty.tools.dotc.core.Contexts._
+import dotty.tools.dotc.core.Definitions._
+import dotty.tools.dotc.core.NameOps._
+import dotty.tools.dotc.core.Names._
+import dotty.tools.dotc.core.StdNames._
+import dotty.tools.dotc.core.Symbols._
+import dotty.tools.dotc.core.Types._
+
+trait ScalaFunction {
+  def arity: Int
+
+  def symbol(implicit ctx: Context): Symbol = defn.FunctionClass(this)
+  def typeRef(implicit ctx: Context): TypeRef = defn.FunctionType(this)
+
+  def isSynthetic: Boolean = ScalaFunction.isSynthetic(arity, isImplicit, isPhantom)
+  def isFunctionXXL: Boolean = false
+  def isImplicit: Boolean
+  def isPhantom: Boolean = false
+
+  def withoutImplicit: ScalaFunction
+  def withoutPhantoms: ScalaFunction = this
+  def erased: ScalaFunction = {
+    val erasedPhantoms = this.withoutPhantoms
+    if (erasedPhantoms.arity > Definitions.MaxImplementedFunctionArity) ScalaFunction.FunctionXXLName
+    else erasedPhantoms.withoutImplicit
+  }
+
+  def name: TypeName
+  def fullName: TypeName = tpnme.scala_ ++ "." ++ name
+
+  def exists: Boolean = true
+
+  override def toString(): String = s"ScalaFunction($name)"
+
+}
+
+object ScalaFunction {
+
+  def apply(sym: Symbol)(implicit ctx: Context): ScalaFunction = apply(defn.scalaClassName(sym))
+
+  def apply(name: TypeName)(implicit ctx: Context): ScalaFunction = {
+    val functionArity = name.functionArity
+    lazy val implicitFunctionArity = name.implicitFunctionArity
+    if (functionArity != -1) {
+      new SimpleFunctionName(functionArity, isImplicit = false)
+    } else if (implicitFunctionArity != -1) {
+      new SimpleFunctionName(implicitFunctionArity, isImplicit = true)
+    } else if (name.phantomFunctionArity != -1) {
+      val phantomicity = new Phantomicity(name.toString.substring(tpnme.PhantomFunction.length))
+      new PhantomFunction(phantomicity, isImplicit = false)
+    } else if (name.implicitPhantomFunctionArity != -1) {
+      val phantomicity = new Phantomicity(name.toString.substring(tpnme.ImplicitPhantomFunction.length))
+      new PhantomFunction(phantomicity, isImplicit = true)
+    } else new NotAFunctionName(name)
```

Comment:
No, it is used to check if the name is a valid function name with`ScalaFunction(name).exists`.